### PR TITLE
[Tests][BWC] add BWC tests to Jenkins

### PR DESCRIPTION
**Pull request by kavilla**
### Description
For OpenSearch and OpenSearch Dashboards. Also fix a bug with the test config
showing incorrectly.

Signed-off-by: Kawika Avilla <kavilla414@gmail.com>
 
### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/705
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]
Can you please add tests for the new jobs? Thank you!

Refer [this](https://github.com/opensearch-project/opensearch-build/blob/main/DEVELOPER_GUIDE.md#tests-for-jenkins-job)

**Comment by kavilla** [comment]
> Can you please add tests for the new jobs? Thank you!
> 
> Refer [this](https://github.com/opensearch-project/opensearch-build/blob/main/DEVELOPER_GUIDE.md#tests-for-jenkins-job)

Could you provide a reference how to handle the dockerAgent variable that is used for the agent. I keep getting something on the lines:
```
    <failure message="groovy.lang.MissingPropertyException: No such property: dockerAgent for class: bwc-test" type="groovy.lang.MissingPropertyException">groovy.lang.MissingPropertyException: No such property: dockerAgent for class: bwc-test
```

I created a binded variable and that doesn't work and tried to return a function. It would appear there is an open issue here: https://github.com/jenkinsci/JenkinsPipelineUnit/issues/507 that actually seems the reason why I'm not able to do this.

**Review by dblock — APPROVED** [review]

**Comment by abhinavGupta16** [comment]
> > Can you please add tests for the new jobs? Thank you!
> > Refer [this](https://github.com/opensearch-project/opensearch-build/blob/main/DEVELOPER_GUIDE.md#tests-for-jenkins-job)
> 
> @abhinavGupta16, could you provide a reference or help with how to handle the dockerAgent variable that is used for the agent. I keep getting something on the lines:
> 
> ```
>     <failure message="groovy.lang.MissingPropertyException: No such property: dockerAgent for class: bwc-test" type="groovy.lang.MissingPropertyException">groovy.lang.MissingPropertyException: No such property: dockerAgent for class: bwc-test
> ```
> 
> I created a binded variable and that doesn't work and tried to return a function. It would appear there is an open issue here: [jenkinsci/JenkinsPipelineUnit#507](https://github.com/jenkinsci/JenkinsPipelineUnit/issues/507) that actually seems the reason why I'm not able to do this.

I don't think we can override the variables declared locally in `script`. Adding a libtester for `detectTestDockerImage()` should help

**Comment by kavilla** [comment]
> I don't think we can override the variables declared locally in `script`. Adding a libtester for `detectTestDockerImage()` should help

Yes I am referring the lib tester for detect test docker image but it still fails here
```
    stage('bwc-test') {
            agent {
                docker {
                    label AGENT_LABEL
                    image dockerAgent.image
                    args dockerAgent.args
                    alwaysPull true
                }
            }
```

**Comment by abhinavGupta16** [comment]
> Yes I am referring the [lib tester for detect test docker image](https://github.com/opensearch-project/opensearch-build/blob/main/tests/jenkins/lib-testers/DetectTestDockerAgentLibTester.groovy) like:
> 
> ```
> this.registerLibTester(new DetectTestDockerAgentLibTester('tests/jenkins/data/opensearch-dashboards-1.2.0-test.yml'))
> ```
> 
> but it still fails here
> 
> ```
>     stage('bwc-test') {
>             agent {
>                 docker {
>                     label AGENT_LABEL
>                     image dockerAgent.image
>                     args dockerAgent.args
>                     alwaysPull true
>                 }
>             }
> ```
> 
> because `No such property: dockerAgent for class: bwc-test`

Ah! I see. I think we should make `dockerAgent` variable globally available or in a parent stage. Currently, it is out of scope for this stage, hence the faliure.

**Comment by abhinavGupta16** [comment]
> > I believe we should make `dockerAgent` variable globally available or in a parent stage.
> 
> Within the test?

In the job

**Comment by kavilla** [comment]
> In the job

Will that create conflicts for every pipeline that uses a variable `dockerAgent `

**Comment by abhinavGupta16** [comment]
> > In the job
> 
> 
> 
> Will that create conflicts for every pipeline that uses a variable `dockerAgent `
> 
> 

It is still being used as a global variable in the job. dockerAgent is declared locally in a stage and within script block. It is then being used in docker block of a different stage. Unfortunately, jenkins allows this, but we shouldn't share local variables between stages

File: jenkins/opensearch-dashboards/bwc-test.jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        )
+        string(
+            name: 'AGENT_LABEL',
+            description: 'The agent label where the tests should be executed, e.g. Jenkins-Agent-al2-x64-c54xlarge-Docker-Host.',
+            trim: true
+        )
+    }
+    stages {
+        stage('verify-parameters') {
+            agent {
```
Comment (kavilla) [inline jenkins/opensearch-dashboards/bwc-test.jenkinsfile:31]: Agent can't be none here due to `fileExists`

Error message from Jenkins : `Required context class hudson.FilePath is missing Perhaps you forgot to surround the code with a step that provides this, such as: node`

**Comment by kavilla** [comment]
@abhinavGupta16, was able to get the tests added. Also got it successfully running in Jenkins as well:

Screen Shot 2022-03-23 at 12 38 34 AM

Results:
```
2022-03-23 07:32:52 INFO     Skipping bwc-tests for opensearch-observability, as it is currently not declared in the test manifest
2022-03-23 07:32:53 INFO     | alerting             | with-security        | PASS  |    0 |
2022-03-23 07:32:53 INFO     | alerting             | without-security     | PASS  |    0 |
2022-03-23 07:32:53 ERROR    | anomaly-detection    | with-security        | FAIL  |    1 |
2022-03-23 07:32:53 ERROR    | anomaly-detection    | without-security     | FAIL  |    1 |
2022-03-23 07:32:53 INFO     | asynchronous-search  | with-security        | PASS  |    0 |
2022-03-23 07:32:53 INFO     | asynchronous-search  | without-security     | PASS  |    0 |
2022-03-23 07:32:53 INFO     | dashboards-reports   | without-security     | PASS  |    0 |
2022-03-23 07:32:53 INFO     | index-management     | with-security        | PASS  |    0 |
2022-03-23 07:32:53 INFO     | index-management     | without-security     | PASS  |    0 |
2022-03-23 07:32:53 ERROR    | k-NN                 | with-security        | FAIL  |    1 |
2022-03-23 07:32:53 ERROR    | k-NN                 | without-security     | FAIL  |    1 |
2022-03-23 07:32:53 ERROR    | ml-commons           | with-security        | FAIL  |    1 |
2022-03-23 07:32:53 ERROR    | ml-commons           | without-security     | FAIL  |    1 |
2022-03-23 07:32:53 INFO     | sql                  | with-security        | PASS  |    0 |
2022-03-23 07:32:53 INFO     | sql                  | without-security     | PASS  |    0 |
[Pipeline] }
```

File: jenkins/opensearch-dashboards/bwc-test.jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,143 @@
+lib = library(identifier: "jenkins@20211118", retriever: legacySCM(scm))
+
```
Comment (abhinavGupta16) [inline jenkins/opensearch-dashboards/bwc-test.jenkinsfile:2]: Can you also add timeouts to both the jobs?

**Review by abhinavGupta16 — COMMENTED** [review]
Thanks for adding the tests.

File: jenkins/opensearch-dashboards/bwc-test.jenkinsfile (hunk 2 above)
Comment (kavilla) [reply jenkins/opensearch-dashboards/bwc-test.jenkinsfile:2]: Updated!

**Review by abhinavGupta16 — APPROVED** [review]

**Pull request merged**